src/compiler/checker.ts

```diff
@@ -8228,6 +8358,27 @@ module ts {
                             break;
                     }
                 }
+
+                if (node.type) {
+                    if (languageVersion >= ScriptTarget.ES6 && isSyntacticallyValidGenerator(node)) {
+                        let returnType = getTypeFromTypeNode(node.type);
+                        if (returnType === voidType) {
+                            error(node.type, Diagnostics.A_generator_cannot_have_a_void_type_annotation);

```

Comment:
But the actual error message you see here would be something about `void` (or `string` or `number` or whatever) missing a `next` method, right? If we're fine reporting that for `string` and `number` then why not `void`. Or maybe I'm misunderstanding what the experience is in the error case.